Source/Dafny/Compilers/Compiler-java.cs

```diff
@@ -19,15 +19,33 @@
 using Bpl = Microsoft.Boogie;
 using static Microsoft.Dafny.ConcreteSyntaxTreeUtils;
 
-namespace Microsoft.Dafny {
-  public class JavaCompiler : Compiler {
-    public JavaCompiler(ErrorReporter reporter)
-      : base(reporter) {
+namespace Microsoft.Dafny.Compilers.Java {
+  public class JavaCompiler : SinglePassCompiler {
+    public override void LateInitialize(ErrorReporter reporter, ReadOnlyCollection<string> otherFileNames) {
+      base.LateInitialize(reporter, otherFileNames);
       IntSelect = ",java.math.BigInteger";
       LambdaExecute = ".apply";
     }
 
-    public override String TargetLanguage => "Java";
+    public override IReadOnlySet<string> SupportedExtensions => new HashSet<string> { ".java" };
+
+    public override string TargetLanguage => "Java";
+    public override string TargetExtension => "java";
+    public override string Basename(string dafnyProgramName) =>
+      TransformToClassName(base.Basename(dafnyProgramName));
+    public override string TargetBaseDir(string baseName) => baseName + "-java";
+    public string TransformToClassName(string baseName) =>
+      Regex.Replace(baseName, "[^_A-Za-z0-9\\$]", "_");
```

Comment:
```suggestion
      Regex.Replace(baseName, "[^_A-Za-z0-9$]", "_");
```
We don't need the backslash because '$' is interpreted literally in a character range, let's fix this while you are refactoring this code.